pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go

```diff
@@ -0,0 +1,387 @@
+		if err = e.recoverEtcdSecrets(ctx); err != nil {
+			return err
+		}
+		e.log.Infoln("Recovered")
+	}
+
+	return nil
+}
+
+func (e *etcdrenew) validateClusterVersion(ctx context.Context) error {
+	e.log.Infoln("validating cluster version now")
+	rawCV, err := e.k.KubeGet(ctx, "ClusterVersion.config.openshift.io", "", "version")
+	if err != nil {
+		return err
+	}
+	cv := &configv1.ClusterVersion{}
+	err = codec.NewDecoderBytes(rawCV, &codec.JsonHandle{}).Decode(cv)
+	if err != nil {
+		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", fmt.Sprintf("failed to decode clusterversion, %s", err.Error()))
+	}
+	clusterVersion, err := version.GetClusterVersion(cv)
+	if err != nil {
+		return err
+	}
+	// ETCD ceritificates are autorotated by the operator when close to expiry for cluster running 4.9+
+	if !clusterVersion.Lt(version.NewVersion(4, 9)) {
+		return api.NewCloudError(http.StatusForbidden, api.CloudErrorCodeForbidden, "", "etcd certificate renewal is not needed for cluster running version 4.9+")
+	}
+	return nil
+}
+
+func fetchNodeNames(ctx context.Context, k adminactions.KubeActions, log *logrus.Entry) ([]string, error) {
+	var masterNodeNames []string
+	var u unstructured.Unstructured
+	var nodes corev1.NodeList
+
+	log.Infoln("fetching node names")
+
+	nodeList, err := k.KubeList(ctx, "node", "")
+	if err != nil {
+		return nil, err
+	}
+	if err = json.Unmarshal(nodeList, &u); err != nil {
+		return nil, err
+	}
+	err = kruntime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &nodes)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, node := range nodes.Items {
+		if _, ok := node.ObjectMeta.Labels["node-role.kubernetes.io/master"]; ok {
+			masterNodeNames = append(masterNodeNames, node.ObjectMeta.Name)
+			continue
+		}
+	}
+	return masterNodeNames, nil
+}
+
+func (e *etcdrenew) validateEtcdOperatorControllersState(ctx context.Context) error {
+	e.log.Infoln("validating etcdOperator Controllers state now")
+	rawEtcd, err := e.k.KubeGet(ctx, "etcd.operator.openshift.io", "", "cluster")
+	if err != nil {
+		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", err.Error())
+	}
+	etcd := &operatorv1.Etcd{}
+	err = codec.NewDecoderBytes(rawEtcd, &codec.JsonHandle{}).Decode(etcd)
+	if err != nil {
+		return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", fmt.Sprintf("failed to decode etcd object, %s", err.Error()))
+	}
+	for _, c := range etcd.Status.Conditions {
+		if _, ok := etcdOperatorControllerConditionsExpected[c.Type]; !ok {
+			continue
+		}
+		if etcdOperatorControllerConditionsExpected[c.Type] != c.Status {
+			return api.NewCloudError(http.StatusInternalServerError, api.CloudErrorCodeInternalServerError, "", "%s is in state %s, quiting.", c.Type, c.Status)
+		}
+	}
+	if e.mode == "renew" {
+		e.lastRevision = etcd.Status.LatestAvailableRevision
+	}
```

Comment:
Please correct me if I'm wrong, but I think this is another case of state creeping into something that could be stateless (and in a function starting with `validate` - something I'd expect to not modify state and only retrieve/read values).

Can we determine `lastRevision` earlier in our reconciliation, perhaps unconditionally? If we do so, then we could use it when/if it's relevant, right? If not, why is it important to only determine last revision _now_ in the code? I'm unclear if so.

Reply:
Made the changes, and introduced a new function to capture the lastRevision. Please review.